### PR TITLE
Install state file

### PR DESCRIPTION
**Pull request by crubier**
**What's the problem this PR addresses?**

https://github.com/yarnpkg/berry/issues/973

**How did you fix it?**

A bit more brutally than in https://github.com/yarnpkg/berry/pull/990

Here we save the whole project state in one big file so there are no resolution to be done at run time.

### CODE REVIEW

**Comment by crubier** [comment]
@arcanis as discussed there https://github.com/yarnpkg/berry/pull/990#issuecomment-591478763

I took a more brutal approach. 

The interesting stuff is in `Project.ts`, in the `generateInstallState` and `restoreInstallState` functions, which simply serialize and restore the whole project state in a file.

However once again I am stuck. I thought this approach would be straightforward enough for me to manage it, but nope! It seems that restoring `accessibleLocators` , `optionalBuilds` , `storedDescriptors` , `storedResolutions ` , `storedPackages ` is not enough ? Or I dont do it right.

**Comment by arcanis** [comment]
When dumping the loaded content I see that `null` (the null) is turned into `null` (the string). It seems this is because we're loading the file through the "failsafe" YAML schema, which doesn't support null as a value (TIL). Try storing it as JSON to see if that goes better.

**Comment by arcanis** [comment]
Also of note: the file is 10M for our repository, I can't imagine what it will be for yours 😬 Maybe some basic format packing optimizations would be a good idea (I think ideally it shouldn't be much more than 1MB for our repo, given that the lockfile is 800KB).

**Comment by crubier** [comment]
It works, and fast!

- On the yarn repo, I get a 300ms resolve time, with a 10Mb `install-state` file (3Mb zipped)
- On the stress test repo, I get a 300ms resolve time, with a 10Mb `install-state` file (3Mb zipped)
- On our private repo, I get a 900ms resolve time, with a 55Mb `install-state` file (12Mb zipped)

Conclusion:

- As expected the `install-state` file is gnarly, but this is not a problem in terms of time performance, even a massive file is fast to generate and restore (less than 1s each for 55Mb), so probably not worth committing to a git repo. 
- This is exactly what we need for our repo, it is now as fast as yarn v1 with this 🎉 Trading a negligible amount of disk space for some **much needed** execution speed.
- However we should not merge this right now, it causes a bug in our repo for one of our scripts (it works fine on like 100 scripts, but 1 fails, I have to investigate but it's probably nothing much) it's a package.bin which is undefined instead of being a string.

But overall: 🎉

**Comment by arcanis** [comment]
> As expected the install-state file is gnarly, but this is not a problem in terms of time performance, even a massive file is fast to generate and restore (less than 1s each for 55Mb), so probably not worth committing to a git repo.

Keep in mind that in the context of zero-install, this file will have to be committed to the repository (although it's possible we can make it an optimization only enabled when the number of virtual packages crosses a threshold). Size is important here.

**Comment by crubier** [comment]
Another possibility could be to not commit this file to the repository, even for zero installs, and use it as a simple cache:

If the file is not present during a `yarn run`, then run the normal resolution algorithm, then write the file. If it is present during a `yarn run`, then load it. I don't know if committing this file can ever be a good idea. The whole point of it would be to be extremely simple to load, not necessitating any resolving at all, just pure loading. 

Of course, I think I can optimize the file from where it is now, I can probably divide its size by 2. But it will never be a "git ideal" file, for my repo it would still be 25Mb so still not git-worthy

**Comment by arcanis** [comment]
Maybe a cache would work, but we'll have to persist the lockfile checksum inside it in order to discard it when switching between branches.

> Of course, I think I can optimize the file from where it is now, I can probably divide its size by 2. But it will never be a "git ideal" file, for my repo it would still be 25Mb so still not git-worthy

We probably can get much more - for example, the lockfile uses `foo: "npm:^1.0.0"` as descriptor where the current state file would stores the hashes (so 256x2 characters) plus the property names, plus ... I think it's worth checking whether deserializing is fast enough that we can afford to do it.

**Comment by arcanis** [comment]
Oh one last idea - if you want to get fancy, maybe using [`v8.serialize`](https://nodejs.org/api/v8.html#v8_v8_serialize_value) together with the compression in [`zlib`](https://nodejs.org/api/zlib.html) would be even faster, as we'd be able to serialize / deserialize the map without even having to iterate over them.

**Comment by crubier** [comment]
@arcanis I love that last approach, it's very much the idea I had in mind. It was super simple to implement as well, and it beat my previous iteration both on speed and file size:

- On our big internal repo the file is 7.6Mb instead of 55Mb. On yarn repo the file is 2.5Mb. 
- And we now match yarn v1 in terms of speed on our big internal repo 🎉 

This is perfect! 

I am now happy with the solution. Now the remaining work is to:

- Maybe see if there is a way to optimize even further. For example in `yarn run`, the slow thing is the `const {project, workspace, locator} = await Project.find(configuration, this.context.cwd);`, taking 1172.026ms, while `restoreInstallState` takes 576.375ms. => @arcanis  is it still needed with this approach, or is there a way to speed up the `setupResolutions` using this?
- Implement the check on the checksum of `yarn.lock` to invalidate the cache if the checksum don't match => @arcanis is there a yarn official way to do this or should I roll my own way to get the checksums?
- I kept the `skipVirtualResolution` parameter that you made in https://github.com/yarnpkg/berry/pull/990 => @arcanis is it still needed in this MR or should I remove these changes ?
- Check what I did already in this MR (the zlib and v8.serialize) => @arcanis did I do it right, or is there a more yarn-kosher way to do what I did?

We are seeing the light at the end of the tunnel!

**Comment by arcanis** [comment]
> is there a way to speed up the `setupResolutions` using `restoreInstallState`?

The time in `Project.find` is likely the one needed to find the configuration, the workspaces, and read their manifests - I'm not sure there's a low-hanging fruit here (at the very least, it should be part of a separate PR).

> Implement the check on the checksum of `yarn.lock` to invalidate the cache if the checksum don't match => @arcanis is there a yarn official way to do this or should I roll my own way to get the checksums?

There isn't at the moment, but I'm fine if you add a private field to `Project` to store the checksum of the lockfile when it's loaded in `setupResolutions`. Oh, and don't forget to hash a constant as well that so that we can force-refresh the cache on future releases with a different cache format.

> I kept the `skipVirtualResolution` parameter that you made in #990 => @arcanis is it still needed in this MR or should I remove these changes ?

Not needed anymore, let's remove it.

> Check what I did already in this MR (the zlib and v8.serialize) => @arcanis did I do it right, or is there a more yarn-kosher way to do what I did?

That seems about right. The code is formatted very differently from the rest of the app so I'll likely make a pass to uniformize it, but the logic itself looks fine.

**Comment by crubier** [comment]
Ok, finished! 

I now write/control the checksum of the lockfile + installStateFileVersion before writing/reading the install state file. It adds 1-3ms depending on the project, so nothing much.

It looks like this 

`this.lockFileChecksum = hashUtils.makeHash(`${INSTALLSTATEFILE_VERSION}`,lockfileContent);`

Let me know what you think @arcanis , change the code style if you like, and let's get this merged!

**Comment by crubier** [comment]
Results of the stress test: The curve looks like this: 

AFTER

Which is much better than it was before: (note the different vertical axis):

BEFORE

**Comment by crubier** [comment]
I had a look at the branch, it looks good to me

File: package.json [hunk 1]
```diff
@@ -58,6 +58,7 @@
     }
   },
   "scripts": {
+    "echo": "echo ok",
```
Comment (merceyz) [inline package.json]:
```suggestion
```

File: packages/yarnpkg-core/package.json [hunk 2]
```diff
@@ -37,6 +37,7 @@
     "@types/got": "^8.3.4",
     "@types/is-ci": "^2.0.0",
     "@types/micromatch": "^3.1.0",
+    "@types/node": "^13.7.6",
```
Comment (merceyz) [inline packages/yarnpkg-core/package.json]:
```suggestion
    "@types/node": "^13.7.0",
```
Should probably use the same version as the rest of the packages

**Comment by arcanis** [comment]
🎉🎉🎉

Thanks @crubier!

**Comment by crubier** [comment]
Thank you so much @arcanis for supporting me through this! My first contribution to yarn, and a lifesaver for my company's tech team, reducing our install and build times by a lot! Best 🎉

**Pull request merged**